### PR TITLE
[Kernel] Support Nvfp4 Cutedsl Moe Swiglu-oai and Relu2(non-gated) Activation

### PR DESCRIPTION
**Pull request by vitamin-chaos**
## Purpose

FlashInfer modified the CuteDSL NVFP4 MoE kernel to support the SwiGLU-OAI activation (https://github.com/flashinfer-ai/flashinfer/pull/3737). The CuteDSL NVFP4 MoE kernel now also supports ReLU² non-gated activation. This PR achieves compatibility with both new activations by passing the activation type and related parameters into the CuteDSL MoE kernel.

## Test Plan

Test input layout prepare: `.venv/bin/python -m pytest tests/kernels/moe/test_flashinfer_cutedsl_layout.py -v`
Test swiglu oai and relu2 accuracy: `.venv/bin/python -m pytest tests/kernels/moe/test_flashinfer_cutedsl_nvfp4_moe.py -v`
Test original model can still run successfully: `.venv/bin/python -m pytest tests/models/quantization/test_nvfp4.py -v -k "flashinfer_cutedsl"`

## Test Result
Test input layout prepare: `.venv/bin/python -m pytest tests/kernels/moe/test_flashinfer_cutedsl_layout.py -v`
```
==================================================================== test session starts =====================================================================
platform linux -- Python 3.12.5, pytest-9.1.1, pluggy-1.6.0 -- /home/scratch.tiekaib_mobile/vllm/.venv/bin/python
cachedir: .pytest_cache
hypothesis profile 'default'
rootdir: /home/scratch.tiekaib_mobile/vllm
configfile: pyproject.toml
plugins: anyio-4.14.1, mock-3.15.1, timeout-2.4.0, buildkite-test-collector-0.1.9, hypothesis-6.155.7, typeguard-4.5.2, asyncio-1.4.0, forked-1.6.0, rerunfailures-16.3, cov-7.1.0, schemathesis-4.21.10, shard-0.1.2
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 3 items
Running 3 items in this shard: tests/kernels/moe/test_flashinfer_cutedsl_layout.py::test_reorder_w13_for_flashinfer_cutedsl_swigluoai_interleaved, tests/kernels/moe/test_flashinfer_cutedsl_layout.py::test_reorder_w13_for_flashinfer_cutedsl_packed_layouts, tests/kernels/moe/test_flashin
fer_cutedsl_layout.py::test_reorder_w13_for_flashinfer_cutedsl_relu2_no_mul_noop

tests/kernels/moe/test_flashinfer_cutedsl_layout.py::test_reorder_w13_for_flashinfer_cutedsl_swigluoai_interleaved PASSED                              [ 33%]
tests/kernels/moe/test_flashinfer_cutedsl_layout.py::test_reorder_w13_for_flashinfer_cutedsl_packed_layouts PASSED                                     [ 66%]
tests/kernels/moe/test_flashinfer_cutedsl_layout.py::test_reorder_w13_for_flashinfer_cutedsl_relu2_no_mul_noop PASSED                                  [100%]

====================================================================== warnings summary ======================================================================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

.venv/lib/python3.12/site-packages/torch/jit/_script.py:365: 14 warnings
  /home/scratch.tiekaib_mobile/vllm/.venv/lib/python3.12/site-packages/torch/jit/_script.py:365: DeprecationWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=============================================================== 3 passed, 16 warnings in 1.43s ===============================================================
```
Test swiglu oai and relu2 accuracy: `.venv/bin/python -m pytest tests/kernels/moe/test_flashinfer_cutedsl_nvfp4_moe.py -v`
```
==================================================================== test session starts =====================================================================
platform linux -- Python 3.12.5, pytest-9.1.1, pluggy-1.6.0 -- /home/scratch.tiekaib_mobile/vllm/.venv/bin/python
cachedir: .pytest_cache
hypothesis profile 'default'
rootdir: /home/scratch.tiekaib_mobile/vllm
configfile: pyproject.toml
plugins: anyio-4.14.1, mock-3.15.1, timeout-2.4.0, buildkite-test-collector-0.1.9, hypothesis-6.155.7, typeguard-4.5.2, asyncio-1.4.0, forked-1.6.0, rerunfailures-16.3, cov-7.1.0, schemathesis-4.21.10, shard-0.1.2
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 2 items
Running 2 items in this shard: tests/kernels/moe/test_flashinfer_cutedsl_nvfp4_moe.py::test_flashinfer_cutedsl_nvfp4_moe_oai_and_relu2[MoEActivation.SWIGLUOAI_UNINTERLEAVE], tests/kernels/moe/test_flashinfer_cutedsl_nvfp4_moe.py::test_flashinfer_cutedsl_nvfp4_moe_oai_and_relu2[MoEActiv
ation.RELU2_NO_MUL]

tests/kernels/moe/test_flashinfer_cutedsl_nvfp4_moe.py::test_flashinfer_cutedsl_nvfp4_moe_oai_and_relu2[MoEActivation.SWIGLUOAI_UNINTERLEAVE] PASSED   [ 50%]
tests/kernels/moe/test_flashinfer_cutedsl_nvfp4_moe.py::test_flashinfer_cutedsl_nvfp4_moe_oai_and_relu2[MoEActivation.RELU2_NO_MUL] PASSED             [100%]

====================================================================== warnings summary ======================================================================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

.venv/lib/python3.12/site-packages/torch/jit/_script.py:365: 14 warnings
  /home/scratch.tiekaib_mobile/vllm/.venv/lib/python3.12/site-packages/torch/jit/_script.py:365: DeprecationWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`.
    warnings.warn(

tests/kernels/moe/test_flashinfer_cutedsl_nvfp4_moe.py: 12 warnings
  /home/scratch.tiekaib_mobile/vllm/.venv/lib/python3.12/site-packages/nvidia_cutlass_dsl/python_packages/cutlass/base_dsl/_mlir_helpers/op.py:121: DeprecationWarning: make_blockscaled_trivial_tiled_mma with ab_dtype is deprecated, use the overload with separate a_dtype and b_dtype ins
tead
    res_or_list = opFunc(*args, **kwargs, loc=loc)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================================================== 2 passed, 28 warnings in 40.69s ===============================================================
sys:1: DeprecationWarning: builtin type swigvarlink has no __module__ attribute
```
Test original model can still run successfully: `.venv/bin/python -m pytest tests/models/quantization/test_nvfp4.py -v -k "flashinfer_cutedsl"`
```
==================================================================== test session starts =====================================================================
platform linux -- Python 3.12.5, pytest-9.1.1, pluggy-1.6.0 -- /home/scratch.tiekaib_mobile/vllm/.venv/bin/python
cachedir: .pytest_cache
hypothesis profile 'default'
rootdir: /home/scratch.tiekaib_mobile/vllm
configfile: pyproject.toml
plugins: anyio-4.14.1, mock-3.15.1, timeout-2.4.0, buildkite-test-collector-0.1.9, hypothesis-6.155.7, typeguard-4.5.2, asyncio-1.4.0, forked-1.6.0, rerunfailures-16.3, cov-7.1.0, schemathesis-4.21.10, shard-0.1.2
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 13 items / 11 deselected / 2 selected
Running 2 items in this shard: tests/models/quantization/test_nvfp4.py::test_nvfp4[flashinfer_cutedsl-True-nvidia/Llama-3.1-8B-Instruct-NVFP4], tests/models/quantization/test_nvfp4.py::test_nvfp4[flashinfer_cutedsl-False-nvidia/Llama-3.1-8B-Instruct-NVFP4]

tests/models/quantization/test_nvfp4.py::test_nvfp4[flashinfer_cutedsl-True-nvidia/Llama-3.1-8B-Instruct-NVFP4] PASSED                                 [ 50%]
tests/models/quantization/test_nvfp4.py::test_nvfp4[flashinfer_cutedsl-False-nvidia/Llama-3.1-8B-Instruct-NVFP4] PASSED                                [100%]

====================================================================== warnings summary ======================================================================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

.venv/lib/python3.12/site-packages/torch/jit/_script.py:365: 14 warnings
  /home/scratch.tiekaib_mobile/vllm/.venv/lib/python3.12/site-packages/torch/jit/_script.py:365: DeprecationWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`.
    warnings.warn(

tests/models/quantization/test_nvfp4.py::test_nvfp4[flashinfer_cutedsl-True-nvidia/Llama-3.1-8B-Instruct-NVFP4]
tests/models/quantization/test_nvfp4.py::test_nvfp4[flashinfer_cutedsl-False-nvidia/Llama-3.1-8B-Instruct-NVFP4]
  /home/utils/Python-3.12.5/lib/python3.12/multiprocessing/popen_fork.py:66: DeprecationWarning: This process (pid=1312091) is multi-threaded, use of fork() may lead to deadlocks in the child.
    self.pid = os.fork()

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================= 2 passed, 11 deselected, 18 warnings in 1543.63s (0:25:43) =================================================
sys:1: DeprecationWarning: builtin type swigvarlink has no __module__ attribute
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by vitamin-chaos** [comment]
https://github.com/flashinfer-ai/flashinfer/pull/3737 was merged. We can test this with flashinfer main branch now.

**Comment by vitamin-chaos** [comment]
> This pull request has merge conflicts that must be resolved before it can be merged. Please rebase the PR, @vitamin-chaos.
> 
> https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork

Done, resolve merge conflict

File: vllm/model_executor/layers/fused_moe/experts/flashinfer_cutedsl_moe.py [hunk 1]
```diff
@@ -56,6 +60,21 @@ def __init__(
...
+        self.gemm1_beta = (
+            float(quant_config.gemm1_beta)
+            if quant_config.gemm1_beta is not None
+            else 1.0
+        )
+        self.gemm1_clamp_limit = (
+            float(quant_config.gemm1_clamp_limit)
+            if quant_config.gemm1_clamp_limit is not None
+            else 7.0
+        )
```
Comment (zyongye) [inline vllm/model_executor/layers/fused_moe/experts/flashinfer_cutedsl_moe.py]: I think we want to set these args at `None` and don't apply them if the parameter in `quant_config` is None, otherwise there will be other model use this file which has these value in default, causing accuracy bug.

**Review by zyongye — APPROVED** [review]
Thanks for doing this. Just tiny changes on how to handle default value!

File: vllm/model_executor/layers/fused_moe/experts/flashinfer_cutedsl_moe.py (hunk 1 above)
Comment (vitamin-chaos) [reply vllm/model_executor/layers/fused_moe/experts/flashinfer_cutedsl_moe.py]: Good point! It won't affect accuracy unless the activation type is swiglu-oai for flashinfer, but it's indeed a little confusing. Modify done.
